### PR TITLE
Dummy implementation of a multi-objective ensemble.

### PR DESCRIPTION
**Pull request by mfeurer**

### CODE REVIEW

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        Index of the model that constitutes the ensemble. This index will
+        be used to select the correct predictions that will be passed during
+        ``fit`` and ``predict``.
+    """
+
+    def __init__(
+        self,
+        task_type: int,
+        metrics: Sequence[Scorer] | Scorer,
+        random_state: int | np.random.RandomState | None,
```
Comment (eddiebergman) [inline autosklearn/ensembles/multiobjective_dummy_ensemble.py:51]: Make sense to have a default = None for this?

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+        # Add some default values -- at least 1 model in ensemble is assumed
+        self.indices_ = [model_index]
+        self.weights_ = [1.0]
+        self.task_type = task_type
+        self.metrics = metrics if isinstance(metrics, Sequence) else [metrics]
+
+    def fit(
+        self,
+        base_models_predictions: np.ndarray | List[np.ndarray],
+        X_data: SUPPORTED_FEAT_TYPES,
```
Comment (eddiebergman) [inline autosklearn/ensembles/multiobjective_dummy_ensemble.py:64]: `X_data` can be None I think? Again, not sure if it should have that as a default

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+            solution=true_targets,
+            prediction=base_models_predictions[self.indices_[0]],
+            task_type=self.task_type,
+            metrics=self.metrics,
+            X_data=X_data,
+        )
+        self.best_model_score_ = loss[self.metrics[0].name]
+        return self
+
+    def predict(self, predictions: Union[np.ndarray, List[np.ndarray]]) -> np.ndarray:
```
Comment (eddiebergman) [inline autosklearn/ensembles/multiobjective_dummy_ensemble.py:111]:
```suggestion
    def predict(self, predictions: np.ndarray | list[np.ndarray]) -> np.ndarray:
```

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+        """
+        return self.best_model_score_
+
+
+class MultiObjectiveDummyEnsemble(AbstractMultiObjectiveEnsemble):
+    def __init__(
+        self,
+        task_type: int,
+        metrics: Sequence[Scorer] | Scorer,
+        random_state: Optional[Union[int, np.random.RandomState]],
```
Comment (eddiebergman) [inline autosklearn/ensembles/multiobjective_dummy_ensemble.py:217]:
```suggestion
        random_state: int | np.random.RandomState | None,
```

Maybe default `= None` too?

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+        backend : Backend
+            Gives access to the backend of Auto-sklearn. Not used.
+        """
+        self.task_type = task_type
+        if isinstance(metrics, Sequence):
+            if len(metrics) == 1:
+                warnings.warn(
+                    "Passed only a single metric to a multi-objective ensemble. "
+                    "Please use a single-objective ensemble in such cases."
+                )
```
Comment (eddiebergman) [inline autosklearn/ensembles/multiobjective_dummy_ensemble.py:244]: Nice catch with this warning :)

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+        # the metric won't be the same as the
+        # one in __main__. we don't use the metric
+        # in the EnsembleSelection so this should
+        # be fine
+        return {key: value for key, value in self.__dict__.items() if key != "metrics"}
+
+    def fit(
+        self,
+        base_models_predictions: List[np.ndarray],
+        X_data: SUPPORTED_FEAT_TYPES,
```
Comment (eddiebergman) [inline autosklearn/ensembles/multiobjective_dummy_ensemble.py:264]: Same here with optional

File: autosklearn/estimators.py [hunk 7]
```diff
@@ -48,7 +51,7 @@ def __init__(
         per_run_time_limit=None,
         initial_configurations_via_metalearning=25,
         ensemble_size: int | None = None,
-        ensemble_class: Type[AbstractEnsemble] | None = EnsembleSelection,
+        ensemble_class: Type[AbstractEnsemble] | Literal["default"] | None = "default",
```
Comment (eddiebergman) [inline autosklearn/estimators.py:54]: Nice, I like this improvement :)

**Review by eddiebergman — APPROVED** [review]
Seems like most of this code I've seen before with all the other ensemble stuff so I didn't look too deep into it. Mostly just nit-picks about typing and defaults.

One thing to use newer typing (renders better in docs, less imports and the recommended way going forward) would be to use `pyupgrade`. Won't block on that though.

```bash
pip install pyupgrade
pyupgrade --py310-plus autosklearn/ensembles/multiobjective_dummy_ensemble.py
```

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py (hunk 2 above)
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: Yes, added this. WRT whether it should have a default, I would say no, using the same argument as above.

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py (hunk 3 above)
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: I applied `pyupgrade`, this should be fixed now.

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py (hunk 4 above)
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: I applied `pyupgrade`, this should be fixed now. I don't think we should add a default here with the same argument as above.

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py (hunk 6 above)
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: Changed this.

File: autosklearn/estimators.py [hunk 8]
```diff
@@ -396,6 +403,9 @@ def __init__(
         self.initial_configurations_via_metalearning = (
             initial_configurations_via_metalearning
         )
+
+        # Need to resolve the ensemble class here so we can act on it below.
+        ensemble_class = self.resolve_ensemble_class(ensemble_class, metric)
```
Comment (eddiebergman) [inline autosklearn/estimators.py:408]: I moved this method into init since it's fairly short and it exposes less methods to the user that they probably don't need to know about.

File: autosklearn/estimators.py [hunk 9]
```diff
@@ -655,19 +665,27 @@ def fit_ensemble(
...
             A metric or list of metrics to score the ensemble with
 
         Returns
         -------
         self
-        """
+        """  # noqa: E501
+
+        # Need to resolve the ensemble class here so we can act on it below.
+        ensemble_class = self.resolve_ensemble_class(ensemble_class, metric)
```
Comment (eddiebergman) [inline autosklearn/estimators.py:687]: Fixed an issue here with regards to what happens when `metric = None`, see the comment in the commits

File: test/test_ensemble_builder/test_multiobjective_dummy_ensemble.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        sklearn.metrics.mean_squared_error,
+        optimum=0,
+        worst_possible_result=-MAXINT,
+        greater_is_better=True,
+        squared=False,
+    )
+
+    ensemble = MultiObjectiveDummyEnsemble(
+        task_type=REGRESSION,
+        metrics=[root_mean_squared_error, negative_root_mean_squared_error],
```
Comment (eddiebergman) [inline test/test_ensemble_builder/test_multiobjective_dummy_ensemble.py:24]: Lol, nice

**Review by eddiebergman — APPROVED** [review]

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        ----------
+        models : dict {identifier : model object}
+            The identifiers are the same as the one presented to the fit()
+            method. Models can be used for nice printing.
+
+        Returns
+        -------
+        list[tuple[float, BasePipeline]]
+        """
+        return self.pareto_set_[0].get_models_with_weights(models)
```
[the remark this answers is not in the document]
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: This is in line with our decision to return an ensemble for the first metric when requesting attributes with respect to individual ensembles to remain a drop-in for Auto-sklearn. Do we need to document this better?

File: autosklearn/estimators.py (hunk 8 above)
Comment (mfeurer) [reply autosklearn/estimators.py]: I'm a bit confused about your comment here. This was in __init__ before your changes and still is in __init__ after your changes.

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 12]
```diff
@@ -57,24 +59,21 @@ def __init__(
             self.metric = [metrics]
         self.random_state = random_state
         self.backend = backend
+        self.pareto_set_: list[SingleModelEnsemble] | None = None
```
Comment (mfeurer) [inline autosklearn/ensembles/multiobjective_dummy_ensemble.py:62]: Hey, is there a reason you added this here? Doesn't this indicate to sklearn that this object has been fitted and therefore it should only be added at fit time?

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+        self.task_type = task_type
+        if isinstance(metrics, Sequence):
+            if len(metrics) == 1:
+                warnings.warn(
+                    "Passed only a single metric to a multi-objective ensemble. "
+                    "Please use a single-objective ensemble in such cases."
+                )
+            self.metrics = metrics
+        else:
+            self.metric = [metrics]
```
[the remark this answers is not in the document]
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py:59]: I would leave it as a warning as this would not stop Auto-sklearn from working. Also, this would only happen if the user willfully passes the `MultiObjectiveDummyEnsemble` to Auto-sklearn. If the user passes `[metric]` to Auto-sklearn without specifying the ensemble model, we would still default to `EnsembleSelection`.

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+
+    def __getstate__(self) -> dict[str, Any]:
+        # Cannot serialize a metric if
+        # it is user defined.
+        # That is, if doing pickle dump
+        # the metric won't be the same as the
+        # one in __main__. we don't use the metric
+        # in the EnsembleSelection so this should
+        # be fine
+        return {key: value for key, value in self.__dict__.items() if key != "metrics"}
```
[the remark this answers is not in the document]
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: We have it in [`EnsembleSelection`](https://github.com/automl/auto-sklearn/blob/0a4dc14ab7cc1f884bc391f101cde91cc2415c86/autosklearn/ensembles/ensemble_selection.py#L95), too. Does this mean we should move this function to `AbstractEnsemble`?

[the remark this answers is not in the document]
Comment (eddiebergman) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: For expediency we can leave it, whenever moving things to automl common, this will get unified.

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (eddiebergman) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: Ahh okay, that makes sense. I didn't realize this is an sklearn interface part. I think the documentation is fine in that case :)

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py (hunk 12 above)
Comment (eddiebergman) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py:62]: So the main reason is for the checkers. In general, setting attributes that are not in `__init__` will lead correctly to mypy error when trying to access it elsewhere as it may not have been set. I was listing out several options but the only one that makes sense is to do:

* Use the `@property pareto_set` everywhere internally and have it do the check it's doing along with `getattr(self, "pareto_set_"). This will raise with a nice error or return the attribute if it exists.

File: autosklearn/estimators.py (hunk 8 above)
Comment (eddiebergman) [reply autosklearn/estimators.py]: This was an outdated comment and reverted with one of the commits #[18b6db4](https://github.com/automl/auto-sklearn/pull/1523/commits/18b6db487698cc21adea9674727117638d5ce41d).

I think the goal was just to remove the method but it didn't make sense to do so in the end.

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: I just unified this nevertheless as this was too easy not to do (and is related to this PR).

File: autosklearn/ensembles/multiobjective_dummy_ensemble.py (hunk 12 above)
Comment (mfeurer) [reply autosklearn/ensembles/multiobjective_dummy_ensemble.py]: Done.

**Review by eddiebergman — APPROVED** [review]
Looks good :)
